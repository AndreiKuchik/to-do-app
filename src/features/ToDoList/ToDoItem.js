import React from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { Draggable } from 'react-beautiful-dnd';
import { deleteToDo, completeToDo } from '../toDoListSlice';

import styles from './ToDoItem.module.css';

const ToDoItem = (props) => {
  const dispatch = useDispatch();

  return (
    <Draggable key={props.id} draggableId={props.id} index={props.index}>
      {(provided) => (
        <li
          ref={provided.innerRef}
          {...provided.draggableProps}
          {...provided.dragHandleProps}
        >
          <input
            type="checkbox"
            checked={props.isCompleted}
            onChange={() => dispatch(completeToDo(props.id))}
            className={styles.isCompletedCheckBox}
          />
          <div
            className={`${styles.toDoText} ${
              props.isCompleted && styles.crossed
            }`}
          >
            {props.text}
          </div>
          <button
            type="button"
            className={styles.deleteButton}
            onClick={() => dispatch(deleteToDo(props.id))}
          >
            Delete
          </button>
        </li>
      )}
    </Draggable>
  );
};

ToDoItem.propTypes = {
  id: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  isCompleted: PropTypes.bool.isRequired,
  index: PropTypes.number.isRequired,
};

export default ToDoItem;
