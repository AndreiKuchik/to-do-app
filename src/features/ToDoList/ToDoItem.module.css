.toDoText {
  display: inline-block;
  margin-left: 1rem;
  color: rgb(125 126 137);
}

.crossed {
  text-decoration: line-through;
  color: rgb(125 126 137);
}

.isCompletedCheckBox {
  width: 1.3em;
  height: 1.3em;
  border-radius: 60%;
  background-color: white;
  vertical-align: middle;
  border: 1px solid #ddd;
  appearance: none;
  outline: none;
  cursor: pointer;
}

.isCompletedCheckBox:checked {
  background-color: #898181;
}

.deleteButton {
  margin-left: auto;
  cursor: pointer;
  padding: 0.5rem;
  border: 1px #373537;
  background-color: #a3a5b2;
  color: white;
  border-radius: 12px;
  margin-right: 1rem;
}
