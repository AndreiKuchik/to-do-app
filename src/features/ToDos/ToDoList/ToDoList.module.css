.toDoList {
  font-family: "Century Gothic", sans-serif;
}

.toDoList ul {
  list-style: none;
  padding-inline-start: 0;
  max-height: 15rem;
  overflow-y: auto;
}

.toDoList ul::-webkit-scrollbar {
  width: 0.3em;
}

.toDoList ul::-webkit-scrollbar-track {
  box-shadow: inset 0 0 6px rgb(238 230 221);
  margin-top: 10px;
  margin-bottom: 10px;
}

.toDoList ul::-webkit-scrollbar-thumb {
  background-color: #36d7b7;
  outline: 1px solid rgb(237, 237, 234);
  border-radius: 8px;
}

.toDoList li {
  border-bottom: 1px solid rgb(222, 219, 219);
  padding: 0.5rem;
  margin: 0 2rem;
  height: 2rem;
  padding: 1rem 0;
  align-items: center;
  display: flex;
}

.todos li:last-child {
  border-bottom: none;
}

.todos h3 {
  font-size: 14px;
  text-align: center;
  color: #a25e5e;
  padding: 0.5rem;
  margin-left: 18%;
  height: 50px;
  padding-top: 20px;
  align-items: center;
  display: flex;
}

.emptyListMessage {
  color: #36d7b7;
  padding: 2rem;
}
