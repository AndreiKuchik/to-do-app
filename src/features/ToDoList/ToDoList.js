import { useSelector, useDispatch } from 'react-redux';
import React, { useEffect } from 'react';
import { DragDropContext, Droppable } from 'react-beautiful-dnd';
import Card from '../../components/Card';

import ToDoFooter from '../ToDoFooter/ToDoFooter';
import { selectToDoList, fetchToDoList, reorderList } from '../toDoListSlice';

import styles from './ToDoList.module.css';
import ToDoItem from './ToDoItem';

const ToDoList = () => {
  const dispatch = useDispatch();
  const toDoList = useSelector(selectToDoList);
  const status = useSelector((state) => state.toDoList.status);

  useEffect(() => {
    if (status === 'idle') {
      dispatch(fetchToDoList());
    }
  }, [status, dispatch]);

  const onDragEnd = (result) => {
    const dragElementId = toDoList[result.source.index].id;
    const dropElementId = toDoList[result.destination.index].id;
    const dragAndDropAction = {
      dragElementId,
      dropElementId,
    };
    dispatch(reorderList(dragAndDropAction));
  };

  return (
    <Card className={styles.toDoList}>
      {toDoList.length !== 0 && (
        <DragDropContext onDragEnd={onDragEnd}>
          <Droppable droppableId="toDoList">
            {(provided) => (
              <ul {...provided.droppableProps} ref={provided.innerRef}>
                {toDoList.map((todo, index) => (
                  <ToDoItem
                    key={todo.id}
                    id={todo.id}
                    text={todo.toDo}
                    index={index}
                    isCompleted={todo.isCompleted}
                  />
                ))}
                {provided.placeholder}
              </ul>
            )}
          </Droppable>
        </DragDropContext>
      )}
      {toDoList.length === 0 && <h3>ToDo list is empty</h3>}
      <ToDoFooter toDosAmount={toDoList.length} />
    </Card>
  );
};

export default ToDoList;
