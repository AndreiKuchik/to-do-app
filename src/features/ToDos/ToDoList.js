/* eslint-disable no-shadow */
import React from 'react';
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd';
import { useAppDispatch, useAppSelector } from '../../core/hooks';
import Card from '../../components/Card';
import { selectToDoList, reorderList } from '../toDoListSlice';
import styles from './ToDoList.module.css';
import ToDoItem from './ToDoItem';

const ToDoList = () => {
  const dispatch = useAppDispatch();
  const toDoList = useAppSelector(selectToDoList);

  const onDragEnd = (result) => {
    const dragElementId = toDoList[result.source.index].id;
    const dropElementId = toDoList[result.destination.index].id;
    const dragAndDropAction = {
      dragElementId,
      dropElementId,
    };
    dispatch(reorderList(dragAndDropAction));
  };

  return (
    <Card className={styles.toDoList}>
      {toDoList.length !== 0 && (
        <DragDropContext onDragEnd={onDragEnd}>
          <Droppable droppableId="toDoList">
            {(provided) => (
              <ul {...provided.droppableProps} ref={provided.innerRef}>
                {toDoList.map((todo, index) => (
                  <Draggable key={todo.id} draggableId={todo.id} index={index}>
                    {(provided) => (
                      <li
                        ref={provided.innerRef}
                        {...provided.draggableProps}
                        {...provided.dragHandleProps}
                      >
                        <ToDoItem
                          key={todo.id}
                          id={todo.id}
                          text={todo.toDo}
                          isCompleted={todo.isCompleted}
                        />
                      </li>
                    )}
                  </Draggable>
                ))}
                {provided.placeholder}
              </ul>
            )}
          </Droppable>
        </DragDropContext>
      )}
      {toDoList.length === 0 && (
        <h3 role="empty-list-text">ToDo list is empty</h3>
      )}
    </Card>
  );
};

export default ToDoList;
