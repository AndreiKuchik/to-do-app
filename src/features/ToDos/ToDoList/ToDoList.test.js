import React from 'react';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import ToDoList from './ToDoList';
import Filters from '../../../common/toDoFiltersEnum';
import Statuses from '../../../common/requestStatusesEnum';

const mockStore = configureStore([]);

describe('Test TodoList component', () => {
  let store;

  it('displays correctly when toDoList is empty', () => {
    // Arrange
    store = mockStore({
      toDoList: {
        toDoList: [],
        status: Statuses.Succeeded,
        filter: Filters.All,
      },
    });

    // Act
    const { container } = render(
      <Provider store={store}>
        <ToDoList />
      </Provider>,
    );
    // Assert
    expect(container).toMatchSnapshot();
  });

  it('displays correctly when there are items in toDoList', () => {
    // Arrange
    store = mockStore({
      toDoList: {
        toDoList: [
          {
            id: '1',
            toDo: 'ToDo1',
            isCompleted: false,
          },
          {
            id: '2',
            toDo: 'ToDo2',
            isCompleted: true,
          },
        ],
        status: Statuses.Succeeded,
        filter: Filters.All,
      },
    });

    // Act
    const { container } = render(
      <Provider store={store}>
        <ToDoList />
      </Provider>,
    );

    // Assert
    expect(container).toMatchSnapshot();
  });

  it('displays correct number of ToDoItems in ToDoList', () => {
    // Arrange
    store = mockStore({
      toDoList: {
        toDoList: [
          {
            id: '1',
            toDo: 'ToDo1',
            isCompleted: false,
          },
          {
            id: '2',
            toDo: 'ToDo2',
            isCompleted: true,
          },
          {
            id: '3',
            toDo: 'ToDo3',
            isCompleted: true,
          },
        ],
        status: Statuses.Succeeded,
        filter: Filters.All,
      },
    });

    // Act
    render(
      <Provider store={store}>
        <ToDoList />
      </Provider>,
    );

    // Assert
    const todoItemsList = screen.getAllByRole('todo-item-checkbox');
    expect(todoItemsList.length).toBe(3);
  });

  it('should show empty list message when todos list is empty', () => {
    // Arrange
    store = mockStore({
      toDoList: {
        toDoList: [],
        status: Statuses.Succeeded,
        filter: Filters.All,
      },
    });

    // Act
    render(
      <Provider store={store}>
        <ToDoList />
      </Provider>,
    );

    // Assert
    const message = screen.getByRole('empty-list-text');
    expect(message).toBeInTheDocument();
  });
});
