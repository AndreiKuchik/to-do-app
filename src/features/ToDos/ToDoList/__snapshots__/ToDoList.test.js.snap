// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TodoList component displays correctly when there are items in toDoList 1`] = `
<div>
  <div
    class="card toDoList"
  >
    <ul
      data-rbd-droppable-context-id="0"
      data-rbd-droppable-id="toDoList"
    >
      <li
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="1"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="1"
        draggable="false"
        role="button"
        tabindex="0"
      >
        <input
          class="toDoCheckbox"
          role="todo-item-checkbox"
          type="checkbox"
        />
        <div
          class="toDoText false"
          role="todo-item-text"
        >
          ToDo1
        </div>
        <button
          class="deleteButton"
          role="todo-item-delete-button"
          type="button"
        >
          Delete
        </button>
      </li>
      <li
        aria-describedby="rbd-hidden-text-0-hidden-text-0"
        data-rbd-drag-handle-context-id="0"
        data-rbd-drag-handle-draggable-id="2"
        data-rbd-draggable-context-id="0"
        data-rbd-draggable-id="2"
        draggable="false"
        role="button"
        tabindex="0"
      >
        <input
          checked=""
          class="toDoCheckbox"
          role="todo-item-checkbox"
          type="checkbox"
        />
        <div
          class="toDoText crossed"
          role="todo-item-text"
        >
          ToDo2
        </div>
        <button
          class="deleteButton"
          role="todo-item-delete-button"
          type="button"
        >
          Delete
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Test TodoList component displays correctly when toDoList is empty 1`] = `
<div>
  <div
    class="card toDoList"
  >
    <h3
      class="emptyListMessage"
      role="empty-list-text"
    >
      ToDo list is empty
    </h3>
  </div>
</div>
`;
